cognite/client/api/assets.py

```diff
@@ -411,3 +454,146 @@ def search(
+            assets_in_response[i].ref_id = asset.ref_id
+            assets_in_response[i].parent_ref_id = asset.parent_ref_id
+        return assets_in_response
+
+    def run(self):
+        while not self.stop:
+            try:
+                request = self.request_queue.get(timeout=0.1)
+            except queue.Empty:
+                continue
+            response = self.client.create(request)
+            assets = self.set_ref_id_on_assets_in_response(request, response)
+            self.response_queue.put(assets)
+
+
+class AssetPoster:
+    def __init__(self, assets: List[Asset], client: AssetsAPI):
+        self.validate_asset_hierarchy(assets)
+
+        self.remaining_assets = [Asset._load(a.dump(camel_case=True)) for a in assets]
+        self.client = client
+
+        self.number_of_assets = len(assets)
+
+        self.ref_id_to_remaining_children = self.initialize_ref_id_to_remaining_children_map(self.remaining_assets)
+        self.ref_id_to_id = {}
+
+        self.request_queue = queue.Queue()
+        self.response_queue = queue.Queue()
+
+        self.created_assets = AssetList([])
+
+    @staticmethod
+    def validate_asset_hierarchy(assets: List[Asset]) -> None:
+        ref_ids = [asset.ref_id for asset in assets if asset.ref_id is not None]
+        ref_ids_set = set(ref_ids)
+        assert sorted(ref_ids) == sorted(list(ref_ids_set)), "Duplicate ref_ids found"
+        for asset in assets:
+            parent_ref = asset.parent_ref_id
+            if parent_ref:
+                assert parent_ref in ref_ids_set, "parent_ref_id '{}' does not point to anything".format(parent_ref)
+                assert asset.parent_id is None, "An asset has both parent_id and parent_ref_id set."
+
+    @staticmethod
+    def initialize_ref_id_to_remaining_children_map(assets: List[Asset]) -> Dict[str, List[Asset]]:
+        ref_id_to_children = {asset.ref_id: [] for asset in assets if asset.ref_id is not None}
+        for asset in assets:
+            if asset.parent_ref_id in ref_id_to_children:
+                ref_id_to_children[asset.parent_ref_id].append(asset)
+        return ref_id_to_children
+
+    def get_descendants_unblocked_by_ref_id(self, assets: List[Asset], limit):
+        unblocked_children = []
+        for asset in assets:
+            if asset.ref_id in self.ref_id_to_remaining_children:
+                asset_children = self.ref_id_to_remaining_children[asset.ref_id].copy()
+                for child_asset in asset_children:
+                    unblocked_children.append(child_asset)
+                    self.ref_id_to_remaining_children[asset.ref_id].remove(child_asset)
+                    if len(unblocked_children) == limit:
+                        return unblocked_children
+
+                    max_descendents = limit - len(unblocked_children)
+                    unblocked_children.extend(self.get_descendants_unblocked_by_ref_id([child_asset], max_descendents))
```

Comment:
Would it be better to add them breadth first? What to we expect to unblock the most assets?